pkg/pillar/cmd/nodeagent/nodeagent.go

```diff
@@ -0,0 +1,484 @@
+	debugPtr := flag.Bool("d", false, "Debug flag")
+	curpartPtr := flag.String("c", "", "Current partition")
+	flag.Parse()
+	debug = *debugPtr
+	debugOverride = debug
+	if debugOverride {
+		log.SetLevel(log.DebugLevel)
+	} else {
+		log.SetLevel(log.InfoLevel)
+	}
+	if *versionPtr {
+		fmt.Printf("%s: %s\n", os.Args[0], Version)
+		return
+	}
+	curpart := *curpartPtr
+	logf, err := agentlog.Init(agentName, curpart)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer logf.Close()
+	if err := pidfile.CheckAndCreatePidfile(agentName); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Infof("Starting %s\n", agentName)
+
+	nodeagentCtx := nodeagentContext{}
+	nodeagentCtx.curPart = strings.TrimSpace(curpart)
+
+	// start the watchdog process timer tick
+	stillRunning := time.NewTicker(25 * time.Second)
+
+	// Make sure we have a GlobalConfig file with defaults
+	types.EnsureGCFile()
+
+	// publisher of NodeAgent Status
+	pubNodeAgentStatus, err := pubsub.Publish(agentName, types.NodeAgentStatus{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	pubNodeAgentStatus.ClearRestarted()
+	nodeagentCtx.pubNodeAgentStatus = pubNodeAgentStatus
+
+	// publisher of Zboot Config
+	pubZbootConfig, err := pubsub.Publish(agentName, types.ZbootConfig{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	pubZbootConfig.ClearRestarted()
+	nodeagentCtx.pubZbootConfig = pubZbootConfig
+
+	// Look for global config such as log levels
+	subGlobalConfig, err := pubsub.Subscribe("", types.GlobalConfig{},
+		false, &nodeagentCtx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	subGlobalConfig.ModifyHandler = handleGlobalConfigModify
+	subGlobalConfig.DeleteHandler = handleGlobalConfigDelete
+	nodeagentCtx.subGlobalConfig = subGlobalConfig
+	subGlobalConfig.Activate()
+
+	// monitor device status through led blinks
+	subLedBlinkConfig, err := pubsub.Subscribe("", types.LedBlinkCounter{},
+		false, &nodeagentCtx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	subLedBlinkConfig.ModifyHandler = handleLedBlinkConfigModify
+	subLedBlinkConfig.DeleteHandler = handleLedBlinkConfigDelete
+	nodeagentCtx.subLedBlinkConfig = subLedBlinkConfig
+	subLedBlinkConfig.Activate()
+
+	// baseline timers
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
+
+	// publish zboot config as of now
+	publishZbootConfigAll(&nodeagentCtx)
+
+	// access the zboot APIs directly, baseosmgr is still not ready
+	nodeagentCtx.updateInprogress = zboot.IsCurrentPartitionStateInProgress()
+	log.Infof("Current partition: %s, inProgress: %v\n", nodeagentCtx.curPart,
+		nodeagentCtx.updateInprogress)
+	log.Infof("configReceivedTime:%v\n", nodeagentCtx.lastConfigReceivedTime)
+
+	// Read the GlobalConfig first
+	// Wait for initial GlobalConfig
+	log.Infof("Waiting for GCInitialized\n")
+	for !nodeagentCtx.GCInitialized {
+		select {
+		case change := <-subGlobalConfig.C:
+			subGlobalConfig.ProcessChange(change)
+
+		case change := <-subLedBlinkConfig.C:
+			subLedBlinkConfig.ProcessChange(change)
+
+		case <-stillRunning.C:
```

Comment:
Please use the new pattern where StillRunning(agentName) is called after the select to handle the lack of fairness in the random select.

Reply:
sure